File: RealWorldLld/mockTest/mock02/scripts/scaffold-leetcode.js
#!/usr/bin/env node
// Scaffolds a LeetCode problem: creates <problemsDir>/<number>/ containing
// <number>.java, <number>.cpp, <number>.js, <number>.py, each with the
// problem statement inserted as a comment block.
//
// Usage: node scaffold-leetcode.js <problemNumber> [problemsDir]

const fs = require("fs");
const path = require("path");
const {
  START_MARKER,
  END_MARKER,
  commentStyleForExt,
  buildCommentBlock,
  fetchProblemByNumber,
} = require("./leetcode-lib");

const LANGUAGE_EXTS = [".java", ".cpp", ".js", ".py"];

function writeCommentBlock(filePath, commentBlock) {
  const existing = fs.existsSync(filePath)
    ? fs.readFileSync(filePath, "utf8")
    : "";

  const blockRegex = new RegExp(
    `[^\\n]*${START_MARKER}[\\s\\S]*?${END_MARKER}\\n?`,
    "m"
  );

  let updated;
  if (blockRegex.test(existing)) {
    updated = existing.replace(blockRegex, `${commentBlock}\n`);
  } else {
    updated = existing.length
      ? `${commentBlock}\n\n${existing}`
      : `${commentBlock}\n`;
  }

  fs.writeFileSync(filePath, updated, "utf8");
}

async function main() {
  const [, , problemNumber, problemsDirArg] = process.argv;

  if (!problemNumber) {
    console.error("Usage: node scaffold-leetcode.js <problemNumber> [problemsDir]");
    process.exit(1);
  }

  const problemsDir =
    problemsDirArg || path.resolve(__dirname, "..");
  const folder = path.join(problemsDir, String(problemNumber));

  fs.mkdirSync(folder, { recursive: true });

  const problem = await fetchProblemByNumber(problemNumber);

  for (const ext of LANGUAGE_EXTS) {
    const filePath = path.join(folder, `${problemNumber}${ext}`);
    const commentStyle = commentStyleForExt(ext);
    const commentBlock = buildCommentBlock(problem, commentStyle);
    writeCommentBlock(filePath, commentBlock);
  }

  console.log(
    `Created ${folder}/ with ${problemNumber}${LANGUAGE_EXTS.join(`, ${problemNumber}`)} for "${problem.title}"`
  );
}

main().catch((err) => {
  console.error(err.message);
  process.exit(1);
});

File: RealWorldLld/mockTest/mock02/scripts/java/ScaffoldLeetcode.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;

/**
 * Scaffolds a LeetCode problem: creates <problemsDir>/<number>/ containing
 * <number>.java, <number>.cpp, <number>.js, <number>.py, each with the
 * problem statement inserted as a comment block.
 *
 * Run from this directory (scripts/java) so the default problemsDir
 * (../.. -> mock02) resolves correctly, or pass a problemsDir explicitly.
 *
 * Usage: java ScaffoldLeetcode <problemNumber> [problemsDir]
 */
public class ScaffoldLeetcode {
    private static final String[] LANGUAGE_EXTS = {".java", ".cpp", ".js", ".py"};

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Usage: java ScaffoldLeetcode <problemNumber> [problemsDir]");
            System.exit(1);
        }

        String problemNumber = args[0];
        Path problemsDir = args.length > 1 ? Path.of(args[1]) : Path.of("../..");
        Path folder = problemsDir.resolve(problemNumber);
        Files.createDirectories(folder);

        Map<String, Object> problem = LeetcodeLib.fetchProblemByNumber(problemNumber);

        StringBuilder created = new StringBuilder();
        for (String ext : LANGUAGE_EXTS) {
            Path filePath = folder.resolve(problemNumber + ext);
            String prefix = LeetcodeLib.commentPrefixForExt(ext);
            String block = LeetcodeLib.buildCommentBlock(problem, prefix);
            LeetcodeLib.writeCommentBlock(filePath, block);

            if (created.length() > 0) created.append(", ");
            created.append(problemNumber).append(ext);
        }

        System.out.println("Created " + folder + "/ with " + created + " for \"" + problem.get("title") + "\"");
    }
}
